src/gotcha_dl.c

```diff
@@ -5,7 +5,124 @@
 #include "elf_ops.h"
 #include <dlfcn.h>
 
+#if defined(__GLIBC__) && __GLIBC__ >= 2 && __GLIBC_MINOR__ < 34
 void* _dl_sym(void* handle, const char* name, void* where);
+#else
+/**
+ * structure used to pass lookup addr and return library address.
+ */
+struct Addrs {
+    ElfW(Addr) library_laddr;
+    ElfW(Addr) lookup_addr;
+};
+/**
+ * This is a callback to get headers for each library.
+ * We check if the caller's virtual address is between base address and the virtual addr + memory size.
+ *
+ * @param info, information about headers
+ * @param size, size of headers
+ * @param data, data from caller
+ * @return
+ */
+int lib_header_callback(struct dl_phdr_info * info, size_t size, void * data) {
+    struct Addrs* addrs = data;
+    ElfW(Addr) end_addr = 0;
+    for (int i=0;i<info->dlpi_phnum;++i) {
+        ElfW(Addr) size = info->dlpi_phdr[i].p_vaddr + info->dlpi_phdr[i].p_memsz;
+        if (end_addr < size) end_addr = size;
+    }
+    ElfW(Addr) base_addr = info->dlpi_addr;
+    ElfW(Addr) end_phdr = base_addr + end_addr;
+    if (base_addr <= addrs->lookup_addr && end_phdr >= addrs->lookup_addr)
+        addrs->library_laddr = base_addr;
+    return 0;
+}
+/**
+ * This function looks for all symbols defined in rela or rel.
+ *
+ * @param name, symbol name
+ * @param lmap, library where we search for symbol.
+ * @return symbol pointer
+ */
+int dl_lookup_symbol_x(const char *name, struct link_map *lmap, void** symbol) {
+    INIT_DYNAMIC(lmap)
+    ElfW(Addr) offset = lmap->l_addr;
+    (void) offset;
+    if (is_rela) {
+        rela = (ElfW(Rela) *) jmprel;
+        for (i = 0; i < rel_count; i++) {
+          ElfW(Addr) offset = rela[i].r_offset;
+          unsigned long symidx = R_SYM(rela[i].r_info);
+          ElfW(Sym) *sym = symtab + symidx;
+          char *symname = strtab + sym->st_name;
+          if (strcmp(symname, name) == 0) {
+            symbol = (void **) (lmap->l_addr + offset);
+            return 0;
+          }
+       }
+    } else {
+        rel = (ElfW(Rel) *) jmprel;
+        for (i = 0; i < rel_count; i++) {
+            ElfW(Addr) offset = rel[i].r_offset;
+            unsigned long symidx = R_SYM(rel[i].r_info);
+            ElfW(Sym) *sym = symtab + symidx;
+            char *symname = strtab + sym->st_name;
+            if (strcmp(symname, name) == 0) {
+                symbol = (void **) (lmap->l_addr + offset);
+                return 0;
+            }
+        }
+    }
+    return -1;
+}
+
+/**
+ * Implement the logic of _dl_sym
+ * 1. find the caller library using the program headers.
+ * 2. find the first library which has the symbol for RTLD_DEFAULT
+ * 3. find the second library which has the symbol for RTLD_NEXT
+ * @param handle, handle for dl operation
+ * @param name, name of the symbol
+ * @param who, the virtual address of the caller
+ * @return symbol pointer
+ */
+
+static void * _dl_sym(void *handle, const char *name, void *who) {
```

Comment:
Let's not treat this as a _dl_sym() replacement. We've reproduced some of the functionality of _dl_sym(RTLD_NEXT, ...), but not all of _dl_sym(). There's going to be cases where this _dl_sym() does the wrong thing, such as IFUNCs, library namespaces, and probably others. 

Instead we should use this new functionality to just replace the library identification part of dlsym(RTLD_NEXT, ...), which gotcha breaks. But we should try to use glib's dlsym() implementation for everything else. So I'd picture the pseudo-code in dlsym_wrapper() that calls this as something like:

```
if called with RTLD_NEXT:
  lib = gotchas_dlsym_rtld_next_lookup() 
  //There may be some more details here in mapping between a 'struct link_map' and the 
  // handle that's passed to dlsym(). I don't remember if dlsym's implementation can take 
  // a link_map pointer.
  return original_dlsym(lib, symbol_name) 
else:
  return original_dlsym(lib, symbol_name)
endif
```

Reply:
renamed function to gotchas_dlsym_rtld_next_lookup

Reply:
We shouldn't have to infinite recurse. This is a wrapped function, and gotcha_get_wrappee() can get you a function pointer to the original_dlsym. By using the original dlsym as much as possible you'll be avoiding a lot of other bugs.

Reply:
I agree now. I checked it and it works. So I changed the code to just do custom lookup for RTLD_NEXT.